crates/shared/src/oneinch_api.rs

```diff
@@ -565,31 +567,31 @@ where
     }
 }
 
+/// A cache for 1Inch API auxiliary data.
 #[derive(Debug, Clone)]
-pub struct ProtocolCache(Arc<Mutex<TimedCache<(), Vec<ProtocolInfo>>>>);
-
-impl ProtocolCache {
-    pub fn new(cache_validity_in_seconds: Duration) -> Self {
-        Self(Arc::new(Mutex::new(TimedCache::with_lifespan_and_refresh(
-            cache_validity_in_seconds.as_secs(),
-            false,
-        ))))
-    }
+pub struct Cache(Arc<Mutex<CacheInner>>);
 
-    pub async fn get_all_protocols(&self, api: &dyn OneInchClient) -> Result<Vec<ProtocolInfo>> {
-        if let Some(cached) = self.0.lock().unwrap().cache_get(&()) {
-            return Ok(cached.clone());
-        }
+#[derive(Debug)]
+pub struct CacheInner {
+    protocols: CacheEntry<Protocols>,
+    spender: CacheEntry<Spender>,
+}
 
-        let all_protocols = api.get_liquidity_sources().await?.protocols;
-        // In the mean time the cache could have already been populated with new
-        // protocols, which we would now overwrite. This is fine.
-        self.0.lock().unwrap().cache_set((), all_protocols.clone());
+#[derive(Debug)]
+struct CacheEntry<T> {
+    store: Option<(T, Instant)>,
+    max_age: Duration,
+}
 
-        Ok(all_protocols)
+impl Cache {
+    pub fn new(cache_validity_in_seconds: Duration) -> Self {
```

Comment:
`in_seconds` doesn't really make sense since you already use `Duration` which is aware of units.
```suggestion
    pub fn new(cache_validity: Duration) -> Self {
```